Cache EDGetToken in InputTag

Dr15Jones opened the pull request:
#### PR description:

Formerly, the InputTag cached the ProductResolverIndex along with all quantities to double check the cached item was being used (in case the same InputTag is being used to retrieve multiple related data products). Now the code caches a EDGetToken and uses the information already available in the EDConsumerBase to verify the correctness of the cached value for the request.

#### PR validation:

All framework unit tests pass.

resolves https://github.com/cms-sw/framework-team/issues/1502

Dr15Jones committed:
Added missing header file

Dr15Jones committed:
Have InputTag cache EDGetToken

Removed direct caching of ProductResolverIndex and related items and replace with EDGetToken.

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"3e423536e5a87ec038420413d5b2da41e8d8f821":{"files":["FWCore/Framework/interface/PrincipalGetAdapter.h"],"squashed":false,"time":1754482933},"ec5043fb9c392d39e3800ada96d02e9092eabff1":{"files":["FWCore/Framework/interface/EDConsumerBase.h","FWCore/Framework/src/EDConsumerBase.cc","FWCore/Framework/src/Principal.cc","FWCore/Utilities/interface/InputTag.h","FWCore/Utilities/src/InputTag.cc","FWCore/Utilities/test/InputTag_t.cpp","FWCore/Utilities/test/test_catch2_InputTag.cc"],"squashed":false,"time":1754483231}},"emoji":{},"last_seen_sha":"ec5043fb9c392d39e3800ada96d02e9092eabff1","signatures":{}} -->

Dr15Jones commented:
The test failures I've looked at all have been caused by `consumes` not being called.  The big question is WHY the old code didn't catch this problem.

Dr15Jones committed:
Added missing consumes calls

cmsbuild commented:
+1

**Size**: This PR adds an extra 80KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ab7515/47564/summary.html
**COMMIT**: 0c4a7bd9673e04d46db4df140c3f5367949c8c59
**CMSSW**: CMSSW_15_1_X_2025-08-06-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/48682/47564/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 2 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-08-06-1100+ab7515/69722/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 8 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-08-06-1100+ab7515/69722/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 4076122
 - DQMHistoTests: [Total failures: 55](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-08-06-1100+ab7515/69722/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4076047
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 215 log files, 184 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-08-06-1100+ab7515/69722/triggerResults): no differences found

Dr15Jones committed:
improvements based on review comments

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48682/45715

cmsbuild commented:
Pull request #48682 was updated. @Dr15Jones, @antoniovagnerini, @atpathak, @cmsbuild, @ctarricone, @makortel, @perrotta, @rseidita, @smuzaffar can you please check and sign again.

makortel commented:
+core

Dr15Jones commented:
@cms-sw/dqm-l2 please sign. The changes are purely technical for the DQM. In fact, the modules I changed didn't actually work correctly ever since we moved to multi-threading at the beginning of Run 2.

perrotta commented:
+alca
- technical, for ShallowDigisProducer

Dr15Jones commented:
@cms-sw/dqm-l2 please sign as the change is only technical for your area.

Dr15Jones commented:
@cms-sw/dqm-l2 again I ask, please sign this change

mandrenguyen commented:
+1
